driver-core/src/test/java/com/datastax/driver/core/AuthenticationTest.java

```diff
@@ -86,4 +89,47 @@ public void should_fail_to_connect_without_credentials() throws InterruptedExcep
         }
     }
 
+    /**
+     * Ensures that authentication is possible even if the server is busy during
+     * SASL handshake.
+     *
+     * @jira_ticket JAVA-1429
+     */
+    @Test(groups = "short")
+    @CCMConfig(dirtiesContext = true)
+    public void should_connect_with_slow_server() throws InterruptedException {
+        Cluster cluster = Cluster.builder()
+                .addContactPoints(getContactPoints())
+                .withPort(ccm().getBinaryPort())
+                .withAuthProvider(new SlowAuthProvider())
+                .withPoolingOptions(new PoolingOptions()
+                        .setHeartbeatIntervalSeconds(1))
+                .build();
+        cluster.connect();
+    }
+
+    private class SlowAuthProvider extends PlainTextAuthProvider {
+
+        public SlowAuthProvider() {
+            super("cassandra", "cassandra");
+        }
+
+        @Override
+        public Authenticator newAuthenticator(InetSocketAddress host, String authenticator) throws AuthenticationException {
+            simulateBusyServer();
```

Comment:
Does a good job reproducing the issue (without the fix) too:

>     18888  [cluster1-nio-worker-0] DEBUG com.datastax.driver.core.Host.STATES - Defuncting Connection[/127.0.1.1:65051-1, inFlight=0, closed=false] because: [/127.0.1.1:65051] Unexpected heartbeat response: ERROR PROTOCOL_ERROR: Unexpected message OPTIONS, expecting SASL_RESPONSE